### PR TITLE
Tree: cleanup typed schema prototype a bit

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

This splits up the typed schema prototype into a few files, and adds a new API layer with optional arguments fort ease of use.

This API is not fully ready for use (ex: APIs for preserving type information when composing SchemaData don't exist), but the parts that are done should work.

See `src/test/feature-libraries/modular-schema/typedSchema/example.spec.ts` for usage.

### CODE REVIEW

File: packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/example.spec.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+});
+
+const rootField = field(optional, [diagramSchema]);
+
+// Collect the schema together.
+// TODO: add APIs for this which preserve the compile time type information.
+const schemaData: SchemaDataAndPolicy = {
+    policy: defaultSchemaPolicy,
+    globalFieldSchema: new Map([[rootFieldKey, rootField]]),
+    treeSchema: new Map([numberSchema, diagramSchema].map((schema) => [schema.name, schema])),
```
Comment (Abe27342) [inline packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/example.spec.ts:58]: should this include `ballSchema`?

File: packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/example.spec.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    name: "Ball",
+    local: {
+        x: field(value, [numberSchema]),
+        y: field(value, [numberSchema]),
+    },
+});
+
+// We can inspect the schema.
+// Note that the inferred type here actually includes the FieldKind's editor,
+// So it would be possible to derive a type safe editing API from this type.
```
Comment (Abe27342) [inline packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/example.spec.ts:36]: is it in a state where you could just demonstrate this with some added lines? might be worth doing so. it's an interesting observation.

**Review by Abe27342 — COMMENTED** [review]
this is a decent amount to grok in one review (I read through the documentation in src/core/schema-stored). I did verify the details here check out as I would expect them (ex: type function mumbo-jumbo) and the examples / type validation tests look reasonable. I'm less confident in saying much about the design choices at a broader scope. Most of the ones that aren't immediately intuitive predate this PR (like local vs global field distinctions, which *do* admittedly have some doc where I mentioned). I left a couple nits but overall this looks good and an improvement to the prototype. Ping me again when you've made changes and I'm OK with approving.

File: packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/typedSchema.spec.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    type check_ = requireAssignableTo<typeof testTreeSchemaFromInfo, typeof testTreeSchema>;
+    type check2_ = requireAssignableTo<typeof testTreeSchema, typeof testTreeSchemaFromInfo>;
+
+    type TestTreeSchema = TypeInfo<typeof testTreeSchema>;
+
+    type _assert = requireTrue<TestTreeSchema["extraGlobalFields"]>;
+
+    type child = FieldInfo<TestTreeSchema["local"][typeof lk1]>;
+
+    // This is an error since this field does not exist:
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/typedSchema.spec.ts]: I did not know about that language feature! Super useful for this kind of thing. Added.

File: packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/example.spec.ts (hunk 1 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/example.spec.ts]: Yes. Fixed.

File: packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/example.spec.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/example.spec.ts:36]: For our real use-cases it needs the TODO down below (to access field content type). Doing that and making a typed API (probably for EditableTree) is the next step.

Comment (CraigMacomber) [reply packages/dds/tree/src/test/feature-libraries/modular-schema/typedSchema/example.spec.ts:36]: I went ahead and tried to make such an example. Doing so uncovered an bunch of issues, but I got it working.

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**